RavenDB-22363 JavaScript indexes: dynamic fields fixes

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22363

### Additional description

This PR contains:
Lucene:
- Create new `IndexField` for dynamic field instead of overriding properties of field from IndexDefinition (runtime)
- - If there is a match between explicit fields and dynamic fields when updating the index definition, it can lead to the replacement process hanging since it would try to create a new dictionary with duplicate keys.
- Preventing overriding indexed term when our definition has enumerable of dynamic fields object with the same name (fix RavenDB-15983 for JS)
- Push dynamic field configuration to the current indexing scope to dynamically retrieve the correct analyzer.
- Ensure backward compatibility for older indexes to match previous behavior.

Corax:
- Update the dynamic field creation flow to match the behavior of C# indexes (update from v5.4).
- Removed persistence of analyzers for dynamic fields
- Allow to override the indexing configuration for dynamic fields when the field name is shared between dynamic and known field
### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [x] Ensured. Please explain how has it been implemented?
- - Lucene fixes are not applied when index version is older.
- [ ] Breaking change
- [] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-22363 Fix javascript dynamic fields in Corax

maciejaszyk committed:
RavenDB-22363 Lucene Dynamic Fields:
- Do not override IndexField definitions in the case of dynamic fields, as it may lead to index replacement hanging (index definition would be changed at runtime).
- Enumerable dynamic field creation should not cause previously indexed terms to be overridden (RavenDB-15983 fix for JS indexes).
- Ensure backward compatibility for older indexes.

arekpalinski commented on src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs:
```diff
@@ -175,12 +175,13 @@ public static class IndexVersion
             public const long PhraseQuerySupportInCoraxIndexes = 60_002;
-            public const long StoreOnlySupportInCoraxIndexes = 60_003;
+            public const long StoreOnlySupportInCoraxIndexes = 60_003; // RavenDB-22369
+            public const long JavaScriptProperlyHandleDynamicFieldsIndexFields = 60_004; //RavenDB-22363
```
Dynamic Fields Index Fields?

maciejaszyk replied:
`Dynamic Fields' Index Fields`?

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexPersistence.cs:
```diff
@@ -24,6 +24,7 @@ namespace Raven.Server.Documents.Indexes.Persistence.Corax;
 
 public sealed class CoraxIndexPersistence : IndexPersistenceBase
 {
+    private const bool DisableDictionaryTraining = false; // [DEBUG ONLY]: disable training.
```
What did you use it for?

maciejaszyk replied:
We've similar thing during index compilation. It is used for disabling training. Pretty useful for debugging indexwriter flow since we do not have then two different index builder (trainer and indexer).

arekpalinski replied:
OK

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/JintCoraxDocumentConverter.cs:
```diff
@@ -312,31 +312,39 @@ private IndexField GetFieldObjectForProcessing(in string propertyAsString, Curre
+#if DEBUG
+            throw new InvalidOperationException($"{nameof(nameProperty)} should not be a null.");
+#endif
+            fieldName = property;
```
Please check `Unreachable code detected` warning here

maciejaszyk replied:
I guess something had to change in Jint itself because this particular code is old and now it seems it cannot be null.

arekpalinski replied:
We have upgraded the Jint meanwhile. You need to verify that.

maciejaszyk replied:
The `GetOwnProperty` in Jint has following description:
https://github.com/sebastienros/jint/blob/65a5ba3afb8d50e128e635a148db0ebe28deba46/Jint/Native/Object/ObjectInstance.cs#L414-L420

I guess instead of comparing `!= null` we have to check it like this: `nameProperty.Value.IsUndefined() == false`

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/JintCoraxDocumentConverter.cs:
```diff
@@ -359,37 +367,49 @@ private static JsValue TryDetectDynamicFieldCreation(string property, ObjectInst
                 if (string.Equals(propertyNameAsString, nameof(CreateFieldOptions.TermVector), StringComparison.OrdinalIgnoreCase))
                 {
-                    field.TermVector = GetEnum<FieldTermVector>(optionValue, propertyNameAsString);
+                    //ignore
```
Can you elaborate on that?

maciejaszyk replied:
This is Corax's JINT converter, TermVector is not used anyway so we can ignore it.

arekpalinski replied:
Maybe a better comment would be nice here

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/JintCoraxDocumentConverter.cs:
```diff
@@ -359,37 +367,49 @@ private static JsValue TryDetectDynamicFieldCreation(string property, ObjectInst
+        if (scope.DynamicFields.TryGetValue(fieldName, out var persistedIndexField))
+        {
+            if (persistedIndexField.Indexing != fieldIndexing)
+                throw new InvalidDataException($"Inconsistent dynamic field creation options were detected. Field '{fieldName}' was created with '{persistedIndexField.Indexing}' analyzer but now '{field.Indexing}' analyzer was specified. This is not supported");
```
Shouldn't we throw only if an index version is `JavaScriptProperlyHandleDynamicFieldsIndexFields` ?

arekpalinski replied:
ping

arekpalinski replied:
@maciejaszyk ?

arekpalinski replied:
Okay, I see `_dynamicFieldsProperlyHandleDynamicConfiguration` check is handling that

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs:
```diff
@@ -171,7 +176,12 @@ void ProcessObject(JsValue valueToInsert, in string propertyAsString, IndexField
+
+                                //We've to wrap dynamic fields to avoid fields cache (and value override), solves: RavenDB_15983
+                                //Also backward compatibility for older indexes.
+                                numberOfCreatedFields = isDynamicFieldsEnumeratorScope && _dynamicFieldsDynamicAnalyzer
+                                    ? GetRegularFields(instance, field, LuceneCreateField(currentIndexingScope, field, CreateValueForIndexing(value, propertyBoost)), indexContext, sourceDocument, out _) 
+                                    : GetRegularFields(instance, field, CreateValueForIndexing(value, propertyBoost), indexContext, sourceDocument, out _);
```
Can we rewrite it to:

```suggestion
object valueForIndexing;

if (isDynamicFieldsEnumeratorScope && _dynamicFieldsDynamicAnalyzer)
    valueForIndexing = LuceneCreateField(currentIndexingScope, field, CreateValueForIndexing(value, propertyBoost));
else
    valueForIndexing = CreateValueForIndexing(value, propertyBoost);
```
?

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs:
```diff
@@ -305,25 +315,31 @@ private static bool TryGetBoostedValue(ObjectInstance valueToCheck, out JsValue
         }
 
-        private static JsValue TryDetectDynamicFieldCreation(string property, ObjectInstance valueAsObject, IndexField field)
+        private JsValue TryDetectDynamicFieldCreation(string property, ObjectInstance valueAsObject, ref IndexField field)
```
Why do you use `ref`? I though the goal is to avoid modifying it completely

maciejaszyk replied:
We're not modifying underlying object inside `IndexDefinition` but value of local `field` variable.  I was thinking about it, but in current flow that would mean doing a lot `out`. I would prefer just working with `ref`.

arekpalinski replied:
Not perfect but we can leave with that.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs:
```diff
@@ -387,5 +418,18 @@ TEnum GetEnum<TEnum>(JsValue optionValue, string propertyName)
+            scope.DynamicFields[field.Name] = field;
+            scope.CreateFieldConverter ??= new LuceneDocumentConverter(scope.Index, new IndexField[] { });
+
+            using var i = scope.CreateFieldConverter.NestedField(scope.CreatedFieldsCount);
```
`i`?

arekpalinski commented:
Please check failing tests

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs:
```diff
@@ -346,32 +362,47 @@ private static JsValue TryDetectDynamicFieldCreation(string property, ObjectInst
                 }
             }
 
+            field = IndexField.Create(fieldName, 
```
do we need to clone the `field` or creating a new one is enough?

maciejaszyk replied:
Field here is local ref variable, we're changing reference to new object. Object inside `IndexDefiniton` is not changed by current flow.

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs:
```diff
@@ -346,32 +362,47 @@ private static JsValue TryDetectDynamicFieldCreation(string property, ObjectInst
+            {
+                //This will get analyzer from dynamic field configuration.
+                CurrentIndexingScope.Current.DynamicFields ??= new();
+                CurrentIndexingScope.Current.DynamicFields[fieldName] = field;
```
this will blindly substitute, I guess it assumes that it was check beforehand? please confirm

maciejaszyk replied:
This is exact copy of current flow inside Lucene C# CreateField:
https://github.com/ravendb/ravendb/blob/7476d9de3a31c1fe66a4af98271a01b08e15e42c/src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs#L493

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Lucene/Documents/JintLuceneDocumentConverter.cs:
```diff
@@ -387,5 +418,18 @@ TEnum GetEnum<TEnum>(JsValue optionValue, string propertyName)
             }
         }
+        
+        private static IEnumerable<AbstractField> LuceneCreateField(CurrentIndexingScope scope, IndexField field, object value)
```
I've seen this code somewhere, was it copied?

maciejaszyk replied:
Yes, this is modified method from C#:
https://github.com/ravendb/ravendb/blob/7476d9de3a31c1fe66a4af98271a01b08e15e42c/src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs#L473-L503

maciejaszyk committed:
RavenDB-22363 Lucene Jint indexes - list of createField object is already flatten.

maciejaszyk committed:
RavenDB-22363 Phrase fields support - check also if field has search as indexing mode.

maciejaszyk commented:
test this in debug please

maciejaszyk committed:
RavenDB-22369
-Removing analyzer persistence tree since we no longer need it.
-Creating virtual IndexedField for dynamic fields that share names with explicit fields.
-When a dynamic field has no configuration during creation, let's check if the explicit field has a configuration before taking the default one.

maciejaszyk commented on src/Corax/Indexing/IndexWriter.cs:
```diff
@@ -131,7 +130,6 @@ private void Init()
             _entryIdToLocation = _transaction.LookupFor<Int64LookupKey>(Constants.IndexWriter.EntryIdToLocationSlice);
             _jsonOperationContext = JsonOperationContext.ShortTermSingleUse();
             _fieldsTree = _transaction.CreateTree(Constants.IndexWriter.FieldsSlice);
-            _persistedDynamicFieldsAnalyzers = _transaction.CreateTree(Constants.IndexWriter.DynamicFieldsAnalyzersSlice);
```
I've removed `_persistedDynamicFieldsAnalyzers` tree since we're not longer needs it. Previously (I mean v5.4) we were storing original term and during deletion we had to analyze it again to retrieve correct term (in meaning the one which is in the `CompactTree`) so we had to store configuration in tree since it may varied for each document. Since we introduced "EntryTable" which contains terms IDs we no longer have original term and deletion happens by term id there is no need to persist it. 

Am I correct? cc @ayende @redknightlois

redknightlois replied:
Sounds about correct, yes.

maciejaszyk commented on src/Corax/Indexing/IndexedField.cs:
```diff
@@ -78,6 +109,13 @@ public override string ToString()
             EntryToTerms = new();
     }
 
+    public IndexedField CreateVirtualIndexedField(IndexFieldBinding dynamicField)
+    {
+        return new IndexedField(dynamicField.FieldId, dynamicField.FieldName, dynamicField.FieldNameLong, dynamicField.FieldNameDouble,
+            dynamicField.FieldTermTotalSumField, dynamicField.Analyzer, dynamicField.FieldIndexingMode, dynamicField.HasSuggestions, dynamicField.ShouldStore,
+            SupportedFeatures, dynamicField.FieldNameForStatistics, FieldRootPage, TermsVectorFieldRootPage, Storage, Textual, Longs, Doubles, this);
+    }
+
```
We had underlying issue when `IndexedField` from known field and dynamic shared the same name. In such scenario, we previously skipped overwritten dynamic field configuration and returned the `IndexedField` from known mapping. We cannot just create duplicate `IndexedField` in such scenario since we're modifying the trees twice in same tx and frequency wouldn't match. I've developed the idea of a `virtual` `IndexedField` for this scenario. Now, all underlying data structures containing terms are now shared between known and dynamic field with the same name. In such scenario we're skipping virtual during commit since all terms from dynamic will be written when committing the known field. This approach is useful for `IndexEntryBuilder`, allowing us to apply custom analyzer/storage configuration, though at the end they're both referencing to the exactly same CompactTree/FST anyway.

cc @redknightlois @ayende

redknightlois replied:
I like it. How that impacts performance?

maciejaszyk replied:
Never measured it, however since I do not created any interface/abstract class and only dealt with it by passing references to other dictionaries I supposed to this change has no performance impact.

arekpalinski commented on src/Corax/Indexing/IndexedField.cs:
```diff
@@ -38,10 +41,11 @@ public struct SpatialEntry
     public readonly bool HasSuggestions;
     public readonly bool ShouldStore;
     public readonly SupportedFeatures SupportedFeatures;
+    public readonly bool IsVirtual;
```
What is the meaning of "virtual"? Does it apply to dynamic fields?

maciejaszyk replied:
Yes, this is strictly for dynamic field but not for every dynamic field. This is basically a shallow copy of known (which we know from compiling/analyzing index definition) field for dynamic which exactly the same name. Idea here is to have reference to in-memory structures which we build to gather all data to insert to the trees since we've two different paths to write to exactly the same trees.

More detail explanation is here: https://github.com/ravendb/ravendb/pull/18545#discussion_r1608054625

arekpalinski reviewed (commented):
Please do the rebase

maciejaszyk committed:
RavenDB-22363 Jint: nameProperty cannot be null now so we've to check underlying value, see more at: https://github.com/sebastienros/jint/blob/65a5ba3afb8d50e128e635a148db0ebe28deba46/Jint/Native/Object/ObjectInstance.cs#L414-L420

maciejaszyk committed:
RavenDB-22363
- Backward compatibility for Corax dynamic fields
- Adding comments

arekpalinski commented:
Can you please solve the conflict?

maciejaszyk commented:
> Can you please solve the conflict?

done

arekpalinski commented:
There is a failing test which might be related to changed made here:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/11082/testReport/junit/SlowTests.Issues/RavenDB_17097/IncludeDocumentDynamicField_options___DatabaseMode___Single___SearchEngineMode___Corax_/

maciejaszyk committed:
RavenDB-22363: Make createField behavior compatible with Lucene version.

arekpalinski commented:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/11132/testReport/junit/SlowTests.Issues/RavenDB_22363/CanUseExplicitAndDynamicFieldsCreationInJavaScriptIndex_options___DatabaseMode___Single___SearchEngineMode___Corax_/
